feat: add `LabelQuestion` and `MultiLabelQuestion` in Python client

alvarobartt opened the pull request:
# Description

This PR adds support to the recently included `LabelSelectionQuestion` and `MultiLabelSelectionQuestion` new types of questions as of #3005 and #3010, respectively, in the Python client side.

Closes #3099 

**Type of change**

- [X] New feature (non-breaking change which adds functionality)

**How Has This Been Tested**

- [x] Add integration tests for the recently included questions

**Checklist**

- [X] I have merged the original branch into my forked branch
- [X] I added relevant documentation
- [X] follows the style guidelines of this project
- [X] I did a self-review of my code
- [ ] I made corresponding changes to the documentation
- [X] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added relevant notes to the CHANGELOG.md file (See https://keepachangelog.com/)

alvarobartt committed:
fix: `required` type-hint

alvarobartt committed:
feat: use `root_validator` & add `Config`

alvarobartt committed:
feat: add `SingleLabelQuestion` and `MultiLabelQuestion`

alvarobartt committed:
fix: add missing `text` field

alvarobartt committed:
feat: add `SingleLabelQuestion` and `MultiLabelQuestion` in `__init__`

alvarobartt committed:
feat: add new questions in `FeedbackDataset`

alvarobartt committed:
fix: include just mandatory fields before API call

alvarobartt committed:
fix: `pytest.raises` string to match

frascuchon commented on src/argilla/client/feedback/dataset.py:
```diff
@@ -199,9 +221,10 @@ def __init__(
         any_required = False
         for question in questions:
-            if not isinstance(question, (TextQuestion, RatingQuestion)):
+            if not isinstance(question, (TextQuestion, RatingQuestion, SingleLabelQuestion, MultiLabelQuestion)):
```
could we use the `AllowedQuestionTypes ` typing somehow?

alvarobartt replied:
We cannot call `isintance` over an `Union`, neither `issubclass` 😢

frascuchon commented on src/argilla/client/feedback/dataset.py:
```diff
@@ -527,7 +560,19 @@ def delete_and_raise_exception(dataset_id: str, exception: Exception) -> None:
+                                "required": True,
+                                "settings": True,
+                            }
+                        ),
```
I think you can use directly a set in this case:

```suggestion
                        question=question.dict(
                            include={ "name", "title", "description", "required", "settings" }
                        ),
```

frascuchon commented on src/argilla/client/feedback/schemas.py:
```diff
@@ -112,40 +117,54 @@ def title_must_have_value(cls, v: Optional[str], values: Dict[str, Any]) -> str:
+class _LabelQuestion(QuestionSchema):
+    settings: Dict[str, Any] = Field({})
+    labels: List[str] = Field(unique_items=True)
+    visible_labels: Optional[PositiveInt] = None
```
We talked about a default value of `20` from the Python client.

alvarobartt replied:
Oh ok I missed that, I just followed the API spec, but 20 makes sense 😄

alvarobartt committed:
fix: set `visible_labels` default value to 20

Co-authored-by: Francisco Aranda <francis@argilla.io>

alvarobartt committed:
fix: use `set` over `dict` in `dict(include=...)`

Co-authored-by: Francisco Aranda <francis@argilla.io>

alvarobartt committed:
fix: allow `List[str]` in `ValueSchema`

As the `MultiLabelQuestion` should log responses as a `List[str]` instead of either `str` or `int`

alvarobartt committed:
refactor: rename `SingleLabelQuestion` to `LabelQuestion`

As discussed internally, we should use `LabelQuestion` and `MultiLabelQuestion` to be more aligned with the API

alvarobartt committed:
feat: add `label_mapping` in `_LabelQuestion`

alvarobartt committed:
fix: set `allow_mutation=False` in `settings`

alvarobartt committed:
fix: `visible_labels` to be `conint(ge=3)`

alvarobartt committed:
fix: `values` & `labels` to have at least 2 items

alvarobartt committed:
feat: remove `label_mapping` in favour of `labels`

alvarobartt committed:
fix: use `get` to get `values`

alvarobartt committed:
fix: skip `root_validator` on field validation failure

alvarobartt committed:
fix: add extra `labels` validation for dicts

alvarobartt committed:
test: `LabelQuestion` & `MultiLabelQuestion` tests

alvarobartt committed:
fix: add `List[str]` as supported type in `FeedbackValueModel`

alvarobartt committed:
fix: `format_as('datasets')` to handle new questions

alvarobartt committed:
test: update unit tests with new questions

gabrielmbmb commented on src/argilla/client/feedback/dataset.py:
```diff
@@ -199,9 +221,10 @@ def __init__(
         any_required = False
         for question in questions:
-            if not isinstance(question, (TextQuestion, RatingQuestion)):
+            if not isinstance(question, (TextQuestion, RatingQuestion, LabelQuestion, MultiLabelQuestion)):
```
Maybe we could create a base class for the questions and just do `isinstance(question, BaseQuestion)`

gabrielmbmb replied:
I've seen that we could use `QuestionSchema`, but if for some reason a user creates a class inheriting from it, it would pass this validation and we would have some errors later.

alvarobartt replied:
We wanted to specifically constraint this to the supported questions in the Python client, that's why :(

gabrielmbmb commented on src/argilla/client/feedback/schemas.py:
```diff
@@ -103,49 +108,80 @@ class QuestionSchema(BaseModel):
+class _LabelQuestion(QuestionSchema):
+    settings: Dict[str, Any] = Field(default_factory=dict, allow_mutation=False)
+    labels: Union[List[str], Dict[str, str]] = Field(unique_items=True, min_items=2)
+    visible_labels: Optional[conint(ge=3)] = 20
```
in the backend we're using `options` instead of `labels`, maybe it's better to align to that here

alvarobartt replied:
Yes I know, but we decided that user-wise it's better to use `labels` as doesn't make much sense to define a `LabelQuestion` with `options` instead of `labels`, but if you both agree @frascuchon, I can revert that to be `options` too 😄

frascuchon commented on src/argilla/client/feedback/schemas.py:
```diff
@@ -103,49 +108,80 @@ class QuestionSchema(BaseModel):
+            if len(set(v.values())) != len(v.values()):
+                raise ValueError("ensure this dict has unique values")
+            return v
+        raise TypeError("ensure this value is a list or a dict")
```
I think this is handled by pydantic using the field definition, isn't it?

alvarobartt replied:
It's not, the `Field(unique_items, min_items)` just works for lists, not for dicts; so we need to add this validation layer besides the `Field`

alvarobartt replied:
Or you mean the `TypeError`?

alvarobartt committed:
fix: remove useless `TypeError`

Co-authored-by: Francisco Aranda <francis@argilla.io>

alvarobartt committed:
fix: use `conlist` instead of `Field`

Co-authored-by: Francisco Aranda <francis@argilla.io>

alvarobartt committed:
fix: replace `raise ValidationError` with `assert`

frascuchon reviewed (approved):
I think we could setup the `settings` attribute as private for the Questions and Field types. Even more, since we're providing the question configuration  through specific properties, these `settings` field  could be generate later when needed (only for the API calls)
